crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::ffi::{c_char, CString};
+
+use crate::exit_codes::{CInputError, ExitCode};
+use num_complex::Complex64;
+use qiskit_accelerate::sparse_observable::{BitTerm, SparseObservable, SparseTermView};
+
+/// @ingroup QkSparseTerm
+/// A term in a [SparseObservable].
+///
+/// This contains the coefficient (``coeff``), the number of qubits of the observable
+/// (``num_qubits``) and pointers to the ``bit_terms`` and ``indices`` arrays, which have
+/// length ``len``. It's the responsibility of the user that the data is coherent,
+/// see also the below section on safety.
+///
+/// # Safety
+///
+/// * ``bit_terms`` must be a non-null, aligned pointer to ``len`` elements of type ``BitTerm``.
+/// * ``indices`` must be a non-null, aligned pointer to ``len`` elements of type ``uint32_t``.
+#[repr(C)]
+pub struct CSparseTerm {
+    coeff: Complex64,
+    len: usize,
+    bit_terms: *mut BitTerm,
+    indices: *mut u32,
+    num_qubits: u32,
+}
+
+impl TryFrom<&CSparseTerm> for SparseTermView<'_> {
+    type Error = CInputError;
+
+    fn try_from(value: &CSparseTerm) -> Result<Self, Self::Error> {
+        check_ptr(value.bit_terms)?;
+        check_ptr(value.indices)?;
+
+        // SAFETY: At this point we know the pointers are non-null and aligned. We rely on C
+        // that the arrays have the appropriate length, which is documented as requirement in the
+        // CSparseTerm class.
+        let bit_terms = unsafe { ::std::slice::from_raw_parts(value.bit_terms, value.len) };
+        let indices = unsafe { ::std::slice::from_raw_parts(value.indices, value.len) };
+
+        Ok(SparseTermView {
+            num_qubits: value.num_qubits,
+            coeff: value.coeff,
+            bit_terms,
+            indices,
+        })
+    }
+}
+
+/// Check the pointer is not null and is aligned.
+fn check_ptr<T>(ptr: *const T) -> Result<(), CInputError> {
+    if ptr.is_null() {
+        return Err(CInputError::NullPointerError);
+    };
+    if !ptr.is_aligned() {
+        return Err(CInputError::AlignmentError);
+    };
+    Ok(())
+}
+
+/// Check a const pointer is non-null and aligned, then cast it to a reference.
+///
+/// # Safety
+///
+/// This function requires ``ptr`` to be non-null, aligned, initialized object of
+/// type ``T``. While the resulting reference exists, the memory pointed to must not
+/// be mutated.
```

Comment:
It's not unsafe to pass a null pointer or a misaligned pointer - we check those and handle them safely.  Worth documenting, but those two bits don't need to be in the safety section.

(Comment repeats wherever this comment does.)

Reply:
Good point 👍🏻

Reply:
Updated in 624c3dd